src/control/cmd/dmg/pretty/network.go

```diff
@@ -40,27 +46,56 @@ func PrintHostFabricMap(hfm control.HostFabricMap, out io.Writer, onlyProviders
 	}
 
 	ew := txtfmt.NewErrWriter(out)
+
 	for _, key := range hfm.Keys() {
+		var hfiMap map[uint32]map[string][]string
+
 		hfs := hfm[key]
 		hosts := control.GetPrintHosts(hfs.HostSet.RangedString(), opts...)
 		lineBreak := strings.Repeat("-", len(hosts))
+		iw := txtfmt.NewIndentWriter(ew, txtfmt.WithPadCount(4))
 		fmt.Fprintf(ew, "%s\n%s\n%s\n", lineBreak, hosts, lineBreak)
-
-		iw := txtfmt.NewIndentWriter(ew, txtfmt.WithPadCount(8))
-		fmt.Fprintf(iw, "Providers: %s\n", strings.Join(hfs.HostFabric.Providers, ","))
 		fmt.Fprintln(ew)
+		fmt.Fprintf(iw, "Available providers: %s\n", strings.Join(hfs.HostFabric.Providers, ", "))
+		fmt.Fprintln(ew)
+
 		if onlyProviders {
 			continue
 		}
 
+		hfiMap = make(map[uint32]map[string][]string)
 		for _, fi := range hfs.HostFabric.Interfaces {
-			fmt.Fprintf(iw, "provider: %s\n", fi.Provider)
-			fmt.Fprintf(iw, "fabric_iface: %s\n", fi.Device)
-			fmt.Fprintf(iw, "pinned_numa_node: %d\n", fi.NumaNode)
-			fmt.Fprintln(ew)
+			if _, ok := hfiMap[fi.NumaNode]; !ok {
+				hfiMap[fi.NumaNode] = make(map[string][]string)
+			}
+			hfiMap[fi.NumaNode][fi.Provider] = append(hfiMap[fi.NumaNode][fi.Provider], fi.Device)
 		}
 
-		fmt.Fprintln(ew)
+		for _, hfi := range hfiMap {
+			for _, devices := range hfi {
+				devices = control.DedupeStringSlice(devices)
```

Comment:
Yeah, this helper shouldn't really be part of the public API. I think with the earlier proposal for adding the `AddInterface()` method to the `control.HostFabric` type you shouldn't need this.

Reply:
This is deduping on a different data structure.  In this case, it is deduping the list of network interfaces that are related to each individual provider.  I'll see if there is some way around it, but I might still need a local dedupe capability if I no longer export that from the control API.

Reply:
Ah, missed that detail. In any case, I'd be fine with a little bit of copying instead of a lot of coupling. It's a simple function. Another option that would be okay would be to add it to the `common` package. There's already a bunch of utility functions there in `common/collection_utils.go`.

Reply:
I think I was concerned about a case that doesn't actually happen.  I am already deduping fabric scan records before they are returned, so the data stream is well conditioned.  At some point while working on this, I thought I saw the potential for the device list to have duplicate data for a given provider, but that doesn't appear to be the case.